### DISCUSSION TITLE
How to Update Boundary Name Dynamically During a Simulation in MOOSE

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#28849] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi everyone,

I’m trying to implement a feature where the boundary name can be updated dynamically during the simulation. Below is an example of the boundary condition I’m working with:

```
[bc_topol0000_LGFlux]
  type = PorousFlowPiecewiseLinearSink
  variable = porepressure_L1
  boundary = 'topol_new'
  pt_vals = '-1e9 1e9'
  multipliers = '-1e9 1e9'
  flux_function = 1e-2
  PT_shift = AuxFx_rhogh
  fluid_phase = 0
  use_mobility = false
  use_relperm = false
[]
```

My goal is to change the boundary name (e.g., from '`topol_new`' to something else) based on simulation time or other control logic.

I have a few ideas but need some guidance on how to implement this:

1. Use [Controls] to update the boundary name dynamically.

2. Create a custom BC class that updates the boundary name based on a list of boundaries and their associated times.
Here’s a starting point of my custom class idea:

```
void PorousFlowBEH_PiecewiseLinearSink::updateBoundary()
{
  Real current_time = _fe_problem.time();  // Get the current simulation time

  for (size_t i = 0; i < _time_per_boundary.size(); ++i)
  {
    if (current_time <= _time_per_boundary[i])
    {
      _current_boundary_index = i;
      break;
    }
  }

  const BoundaryName & active_boundary = _boundaries[_current_boundary_index];

  _console << "Current time: " << current_time << std::endl
           << "Active boundary: " << active_boundary << std::endl
           << std::flush;
}
```

In my new BC block, I’d like to specify multiple boundaries and the time points at which they should change:

```
[bc_topol0000_LGFlux]
  type = PorousFlowBEH_PiecewiseLinearSink
  variable = porepressure_L1
  boundary = 'TOPOLSRF_F000_EQUIL'
  boundaries = 'TOPOLSRF_F000_EQUIL TOPOLSRF_F001_PITX1_DMP1 TOPOLSRF_F002_PITX2_PITF1 TOPOLSRF_F003_PITX3_DMP2'
  time_per_boundary = '0 1 2 3'
  pt_vals = '-1e9 1e9'
  multipliers = '-1e9 1e9'
  flux_function = 1
  char_elem_length_n = charElLength_M0
  PT_shift = AuxFx_rhogh
  fluid_phase = 0
  use_mobility = true
  use_relperm = false
[]
```

Questions:
1. How can I get access to the 'boundary' property within my custom BC class so I can update it dynamically?

2. Is it possible to leverage the [Controls] system to handle this kind of boundary update, or would a new custom class be the better approach?

Thank you!
Traiwit

note: i'm aware of the option that having multiple BC and use multiple  [ConditionalFunctionEnableControl] 

However, we have more than 200+ steps/BCs, we find that having many objects on very large scale simulation is not very optimal

**Comment by GiudGiud** [comment]
Hello

So are you trying to update the boundary name in the mesh? (e.g. the new one does not exist at the beginning)
Or change the boundary on which the BC is active?

> Use [Controls] to update the boundary name dynamically.

That's not currently controllable. To control a parameter, it has to be declared controllable in the validParams. 
Here considering there are conversions done between the names and ids in the initialization of the boundary-restrictable interface, you would need to re-initialize the boundary-restrictable-derived object every time. 

> How can I get access to the 'boundary' property within my custom BC class so I can update it dynamically?

Boundary restriction is currently fixed. 

> However, we have more than 200+ steps/BCs, we find that having many objects on very large scale simulation is not very optimal

It's not optimal but if you are worried about readability, you can stash additional objects in separate files using input files includes. 
Something like:
```
!include all_bcs.i
```

**Comment by Traiwit** [comment, marked as answer]
Hi @GiudGiud 

> So are you trying to update the boundary name in the mesh? (e.g. the new one does not exist at the beginning)
Or change the boundary on which the BC is active?

change the boundary on which the BC is active, this is what I want to do 

noted, i had an idea last night:

use UO to read vector of surface names with respect to time and generate a container surface name

so the surface that feed into BC is always the same (the container surface name), but nodeset/sides will be different at each step.

will give this a try and keep you updated

**Reply by josebastiase** [reply]
Hi Traiwit,

I actually had the same issue a couple of months ago. I ended up writing a Python script to generate a CSV with the active/inactive schedule for the boundary conditions, which I then imported into MOOSE using Functions and Controls. After that, I wrote another script to generate the longer MOOSE input file automatically.

This method worked well for me, but a bit tedious.

**Reply by Traiwit** [reply]
Hi guys,

yeh so i manage to do it, below is the summary of how it works

The `UpdateSurfaceTimeBased` object dynamically switches between predefined boundary surfaces at specific time points during the simulation. It performs a 1:1 copy of the sides from the selected surface to a new "container boundary", clearing any pre-existing sides in the target boundary each time a switch occurs. The copied boundary sides (container boundary) can then be used for applying boundary conditions

where the input file block will look like this:

```
    [./update_surface]
      type = UpdateSurfaceTimeBased
      surface_list = 'TOPOLSRF_F000_EQUIL TOPOLSRF_F001_PITX1_DMP1 TOPOLSRF_F002_PITX2_PITF1 TOPOLSRF_F003_PITX3_DMP2'
      time_per_surface = '0 1 2 3'
      new_boundary_id = 10005                         // container surface ID
      new_boundary_name = topol_time_new  // container surface name
      execute_on = 'TIMESTEP_BEGIN'
    [../]
```

```
#include "UpdateSurfaceTimeBased.h"
#include "MooseMeshUtils.h"
#include "libmesh/mesh.h"

registerMooseObject("moose_behApp", UpdateSurfaceTimeBased);

InputParameters
UpdateSurfaceTimeBased::validParams()
{
  InputParameters params = DomainUserObject::validParams();

  params.addRequiredParam<std::vector<BoundaryName>>("surface_list", 
      "List of surfaces to switch between based on time.");
  params.addRequiredParam<std::vector<Real>>("time_per_surface", 
      "Time points to switch surfaces.");
  params.addRequiredParam<BoundaryName>("new_boundary_name", 
      "The name of the new boundary.");
  params.addRequiredParam<BoundaryID>("new_boundary_id", 
      "The ID of the new boundary.");

  return params;
}

UpdateSurfaceTimeBased::UpdateSurfaceTimeBased(const InputParameters & parameters)
  : DomainUserObject(parameters),
    _pid(_communicator.rank()),
    _neighbor_side(_assembly.neighborSide()),
    _surface_list(getParam<std::vector<BoundaryName>>("surface_list")),
    _time_per_surface(getParam<std::vector<Real>>("time_per_surface")),
    _new_boundary_name(getParam<BoundaryName>("new_boundary_name")),
    _new_boundary_id(getParam<BoundaryID>("new_boundary_id")),
    _boundary_info(_mesh.getMesh().get_boundary_info())
{

}

void UpdateSurfaceTimeBased::initialize()
{
  _remove.clear();
  updateBoundary();  // Select and copy the appropriate boundary based on time
}

void UpdateSurfaceTimeBased::updateBoundary()
{
  Real current_time = _fe_problem.time();
  unsigned int surface_index = 0;

  // Find the correct surface based on the current time
  for (size_t i = 0; i < _time_per_surface.size(); ++i)
  {
    if (current_time <= _time_per_surface[i])
    {
      surface_index = i;
      break;
    }
  }

  const BoundaryName & selected_surface = _surface_list[surface_index];
  const BoundaryID selected_id = MooseMeshUtils::getBoundaryID(selected_surface, _mesh.getMesh());

  // Copy the selected surface to the new boundary
  copyBoundaryData(selected_id, _new_boundary_id);
  _boundary_info.sideset_name(_new_boundary_id) = _new_boundary_name;


  _console << "Selected surface: " << selected_surface << " selected_id: " << selected_id
           << " copied to new boundary: " << _new_boundary_name << " _new_boundary_id: " << _new_boundary_id
           << " at time: " << current_time << std::endl;
}

void UpdateSurfaceTimeBased::copyBoundaryData(BoundaryID from, BoundaryID to)
{
  // _remove.clear();

  _console << " copyBoundaryData selected_id: " << from << " to" << to << std::endl;
  const auto & mesh = _mesh.getMesh();
  const unsigned int n_elem = mesh.n_elem();

  // Clear the target boundary before copying
   // Clear the 'to' boundary before copying from 'from'
  // clearBoundary(to);

  // Iterate over all elements to copy sides from 'from' to 'to'
  for (unsigned int i = 0; i < n_elem; ++i)
  {
    const Elem * elem = mesh.elem_ptr(i);
    for (unsigned int side = 0; side < elem->n_sides(); ++side)
    {
      if (_boundary_info.has_boundary_id(elem, side, from))
      {
        _boundary_info.add_side(elem, side, to);
      }
    }
  }

  // Set the name of the new boundary
  _boundary_info.sideset_name(to) = _new_boundary_name;
}

void UpdateSurfaceTimeBased::finalize()
{
  updateBoundary(); 
  synchronizeBoundaryInfo(_mesh);
}


void 
UpdateSurfaceTimeBased::threadJoin(const UserObject & uo)
{
  const auto & other = static_cast<const UpdateSurfaceTimeBased &>(uo);
  for (const auto & [pid, list] : other._remove)
    _remove[pid].insert(_remove[pid].end(), list.begin(), list.end());
}

void 
UpdateSurfaceTimeBased::synchronizeBoundaryInfo(MooseMesh & mesh)
{
  mesh.getMesh().get_boundary_info().parallel_sync_side_ids();
  mesh.getMesh().get_boundary_info().parallel_sync_node_ids();
  mesh.buildNodeListFromSideList();
}

// void UpdateSurfaceTimeBased::clearBoundary(BoundaryID to)
// {
//   const auto & mesh = _mesh.getMesh();
//   const unsigned int n_elem = mesh.n_elem();

//   // Remove all sides associated with the 'to' boundary
//   for (unsigned int i = 0; i < n_elem; ++i)
//   {
//     const Elem * elem = mesh.elem_ptr(i);
//     for (unsigned int side = 0; side < elem->n_sides(); ++side)
//     {
//       if (_boundary_info.has_boundary_id(elem, side, to))
//       {
//         _boundary_info.remove_side(elem, side, to);
//       }
//     }
//   }

//   // Remove all nodes associated with the 'to' boundary
//   auto & nodeset_map = _boundary_info.get_nodeset_map();
//   std::vector<const Node *> nodes_to_remove;
//   for (const auto & pair : nodeset_map)
//   {
//     if (pair.second == to)
//       nodes_to_remove.push_back(pair.first);
//   }

//   for (const auto node : nodes_to_remove)
//     _boundary_info.remove_node(node, to);
// }
```

**Reply by GiudGiud** [reply]
I think you will want your objects to be boundary restricted to these new boundaries (as well as the old one?) from the very start.

**Reply by Traiwit** [reply]
> to be boundary restricted to these new boundaries

Hi @GiudGiud i dont get this, could you please explain?

Thank you!

**Reply by GiudGiud** [reply]
While you can change boundaries dynamically, you cannot change the boundary restriction of an object. 

So either you already set the boundary restriction to work for the whole simulation, or you use multiple objects and turn some on/off as the boundaries change

**Reply by Traiwit** [reply]
Hi @GiudGiud, I tested it, it kinda works fine (for now), 

so below is my BC, while the boundary name (`topol_time_new`) stays the same, the nodes and sides inside this boundary can change dynamically due to `UpdateSurfaceTimeBased`.

I guess, this is the first case you mentioned.

```
 [bc_topol0000_LGFlux]
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure_L1
    boundary = 'topol_time_new'
    pt_vals = '-1e9 1e9'
    multipliers = '-1e9 1e9'
    flux_function = 1
    PT_shift = -98100   # in Pa, ie., -9810.N/m3*-10m
    fluid_phase = 0
    use_mobility = true   # default is false
    use_relperm = false    # default is false
  []
```

**Reply by GiudGiud** [reply]
if the name / id stays the same this should work

**Reply by GiudGiud** [reply]
if the name / id stays the same this should work

**Comment by Yulongggggg** [comment]
Hi Traiwit, I am also a mining engineering PhD student using MOOSE to simulate the tunneling process, can I ask you some questions regarding MOOSE recently, I have been working on simulating the excavation process in plasticity , may I have your email?

**Reply by Traiwit** [reply]
Hi @Yulongggggg 

For mech simulation we still use Abaqus since MOOSE explicit is still much slower

but if you have any question re: MOOSE hydro-mining related, please feel free to tag me.

note: for hydro-mech we do 2 ways coupling between MOOSE and Abaqus.

**Reply by GiudGiud** [reply]
We have a new time integrator for explicit maybe that will help enough?
It's called DirectCentralDifference
What have you been using for time integration?

This might be minor I dont know

**Reply by Traiwit** [reply]
Hi @GiudGiud 

Thanks, originally I used `CentralDifference`, with `solve_type=lumped`

I will check this out and give you an update on performance 

been giving up on this a couple times already
